### PR TITLE
Fides Pydantic V2 Upgrade

### PR DESCRIPTION
**Pull request by pattisdr**
Closes #PROD-2263

❗ Dependent on https://github.com/ethyca/fideslang/pull/11
ℹ️ Replaces https://github.com/ethyca/fides/pull/4442

### Description Of Changes

Pydantic V2/Fast API Upgrade

### Code Changes

- Removes support for Python 3.8-
-  Upgrades Pydantic to V2, upgrades FastAPI, and other supporting dependencies
- class Config -> `model_config=ConfigDict(...)`
- `dict()` -> `model_dump()`, often with mode="json" since model_dump() is not always serializable as json, depending on model contents
- `parse_obj()` -> `model_validate()`
- `from_orm()` -> `model_validate()`
- conlist -> Annotated[List[], Field(max_length=50)]
- `root_validators` -> `model_validators`. mode="after" is often the better way to go because before validators can be called in surprising places, where sometimes the incoming data is a dict, or other times, the schema itself
- `__init_subclass__` -> `__pydantic_init_subclass__`
- `__fields__` -> `model_fields()`
- `customise_sources` -> `customise_setting_sources`
- `update_forward_refs` -> `model_rebuild`
-  Always=True -> `validate_default=True` on the field directly
- `ConstrainedStr` now need to be defined using StringConstraints
- Sometimes our str(errors) needed to be wrapped with jsonable_encoder.  Errors are not always json serializable
- Lots of default values of `None` added to keep Optional fields optional
- `@app.on_event("startup")` is deprecated, using [Lifespan](https://fastapi.tiangolo.com/advanced/events/) instead
- Updated definitions of custom types: SafeStr, HtmlStr, PhoneNumber, GPPMechanismConsentValue, URLOrigin, and CssStr
- Pydantic V2 can't serialize unknown types. Adding FieldSerializers on some of our Python classes that we're using as types for DSR purposes since Pydantic V2 can't serialize them otherwise.
- Using duck typing to serialize deeping nested collections to match original behavior for DSR processing.  A customer's collection might not be fully defined with a Pydantic schema which was causing some of the collection representation to get dropped in V2.
- Updates to `NoValidationSchema` so it can still be used as a Pydantic schema that skips validation allowing a schema to be used for docs-only
- You can't just stash random things on schemas, use `json_schema_extra`
- We return a lot of Pydantic error messages in failed API calls, and the text has changed
- Models are no longer equal to the dicts containing their data, resulting in a lot of test changes
- Middleware can no longer be updated after the application has started - hack to get this working since we allow cors origins to be updated after the fact
- Exposes admin-ui settings in the Config API 

### Steps to Confirm

This is something that really needs a full regression.  These are features that needed more attention:

* [x] Checking adding multiple types of integrations, postgres, bigquery, attentive, mailchimp
* [x] Verify connection secret schemas show up in docs
* [ ] Running DSR's with Postgres and Mongo and verifying deeply nested results
* [x] Verifying saving messaging secrets
* [x] Verifying saving storage secrets
* [x] General verification that secrets aren't leaked in error messages/logs.  Pydantic error messages are more verbose.
* [x] CORS origins, updating CORS origins after the application is running
* [x] Setting env vars in the toml and as env vars
* [x] CLI commands

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] if there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated
* If there are any database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!

### CODE REVIEW

**Comment by ThomasLaPiana** [comment]
@pattisdr You're a rockstar! :laughing:

**Comment by pattisdr** [comment]
Your work has been so helpful @ThomasLaPiana especially seeing how you resolved some of the more subjective pieces. More to do but making progress!

**Comment by ThomasLaPiana** [comment]
> Your work has been so helpful @ThomasLaPiana especially seeing how you resolved some of the more subjective pieces. More to do but making progress!

I'm working on moving one of our services from Python 3.7 to 3.11 and including the Pydantic upgrade...I will also be referring to this work! :smile:

File: tests/ops/models/test_application_config.py
Comment (pattisdr) [inline tests/ops/models/test_application_config.py:224]: `model_dump_json` stores in a more condensed format, to save space, so I'm adjusting the test

File: src/fides/api/task/create_request_tasks.py
Comment (pattisdr) [inline src/fides/api/task/create_request_tasks.py:191]: Another important place for DSR 3.0 - to save the collection on the RequestTask with duck typing so all the nested sub fields are persisted

File: src/fides/api/custom_types.py
Comment (pattisdr) [inline src/fides/api/custom_types.py:127]: AnyURL's add a trailing slash to everything now, and are used for cors_origins, we can't have trailing slashes there.  Updated the validator to strip it off.

**Comment by pattisdr** [comment]
This gets out of date quickly, will reconcile within the next couple of days, focused on Plus at the moment -

File: src/fides/config/utils.py
Comment (pattisdr) [inline src/fides/config/utils.py:69]: @galvana I think we meant to expose this here, but double checking that it is safe to show

@daveqnet this is adding admin ui settings to the `CONFIG_KEY_ALLOWLIST`

Comment (galvana) [reply src/fides/config/utils.py:69]: Yes! This is safe, it's just the Admin UI URL

File: src/fides/api/schemas/connection_configuration/connection_secrets_bigquery.py
Comment (galvana) [inline src/fides/api/schemas/connection_configuration/connection_secrets_bigquery.py:23]: Why do we need to add the `default` keyword here? Is this something we should change for all usages of `Field`?

Comment (pattisdr) [reply src/fides/api/schemas/connection_configuration/connection_secrets_bigquery.py:23]: we don't have to add it, it's not required but doesn't hurt -

File: src/fides/config/utils.py
Comment (pattisdr) [reply src/fides/config/utils.py:69]: thanks for confirming 👍

File: src/fides/api/schemas/privacy_request.py
Comment (galvana) [inline src/fides/api/schemas/privacy_request.py:129]: So clean! ✨

File: src/fides/connectors/aws.py
Comment (galvana) [inline src/fides/connectors/aws.py:247]: Was this just something we missed in the original implementation? Returning a set of tuples instead of a dictionary?

File: src/fides/api/db/samples.py
Comment (galvana) [inline src/fides/api/db/samples.py:123]: What was the reason for this change?

File: src/fides/api/schemas/connection_configuration/connection_secrets_bigquery.py
Comment (galvana) [reply src/fides/api/schemas/connection_configuration/connection_secrets_bigquery.py:23]: Got it! Thanks

File: src/fides/connectors/aws.py
Comment (pattisdr) [reply src/fides/connectors/aws.py:247]: This was something I updated after a merge with main yesterday, it's brand new, I think it might have been getting coerced with pydantic v1, and not coerced in Pydantic v2.

File: src/fides/api/db/samples.py
Comment (pattisdr) [reply src/fides/api/db/samples.py:123]: mypy I believe -

Comment (pattisdr) [reply src/fides/api/db/samples.py:123]: upgrading mypy flagged hundreds of places -

File: requirements.txt
Comment (pattisdr) [reply requirements.txt]: `bump-pydantic` has been removed

**Review by galvana — APPROVED** [review]
Thanks for tracking the serialization warnings! I'm going to approve to unblock this PR but let me know if there is any other testing I should do before we finally merge.

File: clients/admin-ui/src/types/api/models/ClassifyParams.ts
Comment (eastandwestwind) [inline clients/admin-ui/src/types/api/models/ClassifyParams.ts:11]: why does this return only null?

Comment (eastandwestwind) [reply clients/admin-ui/src/types/api/models/ClassifyParams.ts:11]: Hmm, ok, I'll update back to using `Array<any>` for now.

File: clients/admin-ui/src/types/api/models/ConnectionConfigurationResponse.ts
Comment (eastandwestwind) [inline clients/admin-ui/src/types/api/models/ConnectionConfigurationResponse.ts:25]: here's another instance of pydantic using `null`. I'm changing this back to `any` for now

File: clients/admin-ui/src/types/api/models/ClassifyParams.ts
Comment (pattisdr) [reply clients/admin-ui/src/types/api/models/ClassifyParams.ts:11]: @eastandwestwind did I accidentally override this? Should I put back to Array<any>?

Comment (eastandwestwind) [reply clients/admin-ui/src/types/api/models/ClassifyParams.ts:11]: If the tests pass, then this is fine as `null`. I don't want to run into issues in the future where every time we autogenerate types, we have to do manual changes like this.

I think the real question is how we get pydantic to better transform types like `Optional[Iterable]` from the BE.

File: clients/admin-ui/src/types/api/models/ConnectionConfigurationResponse.ts
Comment (eastandwestwind) [reply clients/admin-ui/src/types/api/models/ConnectionConfigurationResponse.ts:25]: keep as `null` if tests pass

File: clients/admin-ui/src/types/api/models/ClassifyParams.ts
Comment (pattisdr) [reply clients/admin-ui/src/types/api/models/ClassifyParams.ts:11]: Thank you 🙏

File: clients/admin-ui/src/types/api/models/ConnectionConfigurationResponse.ts
Comment (pattisdr) [reply clients/admin-ui/src/types/api/models/ConnectionConfigurationResponse.ts:25]: got it 👍

**Comment by ThomasLaPiana** [comment]
@pattisdr bravo!

**Pull request merged**
